src/common/blkdev.cc

```diff
@@ -544,34 +574,129 @@ bool get_vdo_utilization(int fd, uint64_t *total, uint64_t *avail)
 }
 
 #elif defined(__FreeBSD__)
-#include <sys/disk.h>
 
-int get_block_device_size(int fd, int64_t *psize)
+const char *BlkDev::sysfsdir() const {
+  assert(false);  // Should never be called on FreeBSD
+  return "";
+}
+
+int BlkDev::dev(char *dev, size_t max) const
+{
+  struct stat sb;
+
+  if (fstat(fd, &sb) < 0)
+    return -errno;
+
+  snprintf(dev, max, "%" PRIu64, (uint64_t)sb.st_rdev);
+
+  return 0;
+}
+
+int BlkDev::get_size(int64_t *psize) const
 {
   int ret = ::ioctl(fd, DIOCGMEDIASIZE, psize);
   if (ret < 0)
     ret = -errno;
   return ret;
 }
 
-bool block_device_support_discard(const char *devname)
+bool BlkDev::support_discard() const
 {
-  return false;
+  struct diocgattr_arg arg;
+
+  strlcpy(arg.name, "GEOM::candelete", sizeof(arg.name));
+  arg.len = sizeof(arg.value.i);
+  if (ioctl(fd, DIOCGATTR, &arg) == 0) {
+    return (arg.value.i != 0);
+  } else {
+    return false;
+  }
```

Comment:
nit: Not sure there is much point in implementing `BlkDev::support_discard` while `BlkDev::discard` is not implemented. It looks like currently is used only in os/filestore/FileJournal.cc. And if `support_discard()` will return true, then `discard` will be called and will just generate debug(1) error messages.

But I don't mind leaving it as is.

Reply:
good point.  i will guard this part with `FREEBSD_WITH_TRIM`. probably we will be able to figure out how to trim the block device on FreeBSD.